src/coreComponents/physicsSolvers/wavePropagation/AcousticFirstOrderWaveEquationSEMKernel.hpp

```diff
@@ -321,6 +321,355 @@ struct DampingMatrixKernel
 
 };
 
+template< typename FE_TYPE >
+struct ComputeTimeStep
+{
+
+  ComputeTimeStep( FE_TYPE const & finiteElement )
+    : m_finiteElement( finiteElement )
+  {}
+
+  /**
+   * @brief Compute timestep using power iteration method
+   */
+  template< typename EXEC_POLICY, typename ATOMIC_POLICY >
+  real64
+  launch( localIndex const sizeElem,
+          localIndex const sizeNode,
+          arrayView2d< WaveSolverBase::wsCoordType const, nodes::REFERENCE_POSITION_USD > const X,
+          arrayView2d< localIndex const, cells::NODE_MAP_USD > const & elemsToNodes,
+          arrayView1d< real32 const > const density,
+          arrayView1d< real32 > const mass )
+  {
+
+    constexpr localIndex numNodesPerElem = FE_TYPE::numNodes;
+    constexpr localIndex numQuadraturePointsPerElem = FE_TYPE::numQuadraturePoints;
+
+    real64 const epsilon = 0.00001;
+    localIndex const nIterMax = 10000;
+    localIndex numberIter = 0;
+    localIndex counter = 0;
+    real64 lambdaNew = 0.0;
+
+    array1d< real32 > const p( sizeNode );
+    array1d< real32 > const pAux( sizeNode );
+    array2d< real32 > const ux( sizeElem, numNodesPerElem );
+    array2d< real32 > const uy( sizeElem, numNodesPerElem );
+    array2d< real32 > const uz( sizeElem, numNodesPerElem );
+
+    arrayView1d< real32 > const pView = p;
+    arrayView1d< real32 > const pAuxView = pAux;
+    arrayView2d< real32 > const uxView = ux;
+    arrayView2d< real32 > const uyView = uy;
+    arrayView2d< real32 > const uzView = uz;
+
+
+    //Randomize p values
+    srand( time( NULL ));
+    for( localIndex a = 0; a < sizeNode; ++a )
+    {
+      pView[a] = (real64)rand()/(real64) RAND_MAX;
+    }
+
+    //Step 1: Normalize randomized pressure
+    real64 normP= 0.0;
+    WaveSolverUtils::dotProduct( sizeNode, pView, pView, normP );
+
+    forAll< EXEC_POLICY >( sizeNode, [=] GEOS_HOST_DEVICE ( localIndex const a )
```

Comment:
Can't you use the stiffness kernel directly to compute K.x instead of adding all this code? It seems like tere should be a lot of overlap with the usual computation of a forward step that the kernel already does...